numpy/__init__.pyi

```diff
@@ -349,15 +316,112 @@ class ndarray(Iterable, Sized, SupportsInt, SupportsFloat, SupportsComplex,
         def __matmul__(self, other): ...
         def __rmatmul__(self, other): ...
 
-    def __neg__(self) -> ndarray: ...
-    def __pos__(self) -> ndarray: ...
-    def __abs__(self) -> ndarray: ...
-    def __invert__(self) -> ndarray: ...
+    def __neg__(self: _ArraySelf) -> _ArraySelf: ...
+    def __pos__(self: _ArraySelf) -> _ArraySelf: ...
+    def __abs__(self: _ArraySelf) -> _ArraySelf: ...
+    def __invert__(self: _ArraySelf) -> _ArraySelf: ...
 
     # TODO(shoyer): remove when all methods are defined
     def __getattr__(self, name) -> Any: ...
 
 
+class ndarray(_ArrayLike, Iterable, Sized):
+    real: ndarray
+    imag: ndarray
+
+    @property
+    def dtype(self) -> _Dtype: ...
+    @dtype.setter
+    def dtype(self, value: _DtypeLike): ...
+
+    @property
+    def ctypes(self) -> _ctypes: ...
+
+    @property
+    def shape(self) -> _Shape: ...
+    @shape.setter
+    def shape(self, value: _ShapeLike): ...
+
+    @property
+    def flat(self) -> flatiter: ...
+
+    @property
+    def strides(self) -> _Shape: ...
+    @strides.setter
+    def strides(self, value: _ShapeLike): ...
+
+    # Many of these special methods are irrelevant currently, since protocols
+    # aren't supported yet. That said, I'm adding them for completeness.
+    # https://docs.python.org/3/reference/datamodel.html
+    def __len__(self) -> int: ...
+    def __getitem__(self, key) -> Any: ...
+    def __setitem__(self, key, value): ...
+    def __iter__(self) -> Any: ...
+    def __contains__(self, key) -> bool: ...
+    def __index__(self) -> int: ...
+
+class generic(_ArrayLike):
+    @property
+    def base(self) -> None: ...
+class _is_real(generic):
+    @property
+    def real(self: _ArraySelf) -> _ArraySelf: ...
+
+    @property
+    def imag(self: _ArraySelf) -> _ArraySelf: ...
+
+class number(generic): ...
+class bool_(_is_real): ...
+class object_(generic): ...
+class datetime64(_is_real): ...
+
+class integer(_is_real): ...
+class signedinteger(integer): ...
+class int8(signedinteger): ...
+class int16(signedinteger): ...
+class int32(signedinteger): ...
+class int64(signedinteger): ...
+class timedelta64(signedinteger): ...
+
+class unsignedinteger(integer): ...
+class uint8(unsignedinteger): ...
+class uint16(unsignedinteger): ...
+class uint32(unsignedinteger): ...
+class uint64(unsignedinteger): ...
+
+class inexact(number): ...
+class floating(_is_real): ...
+class float16(floating): ...
+class float32(floating): ...
+class float64(floating): ...
+
+class complexfloating(inexact): ...
+class complex64(complexfloating):
+    @property
+    def real(self) -> float32: ...
+    @property
+    def imag(self) -> float32: ...
+class complex128(complexfloating):
+    @property
+    def real(self) -> float64: ...
+    @property
+    def imag(self) -> float64: ...
+
+class flexible(_is_real): ...
+class void(flexible): ...
+class character(_is_real): ...
```

Comment:
This is correct, but it's also bizarre:
```
>>> np.bytes_(b'test').real
'test'
>>> np.bytes_(b'test').imag
''
```

What were we thinking here?

Reply:
If I had to guess, I think `imag` by default returns the "0" version of the data type -- while this makes sense for numbers, it has bizarre consequences for things that aren't numbers:

```python
In [4]: np.datetime64(dt.datetime.now()).real
Out[4]: numpy.datetime64('2018-03-26T15:22:14.672804')

In [5]: np.datetime64(dt.datetime.now()).imag
Out[5]: numpy.datetime64('1970-01-01T00:00:00.000000')
```

or 

```python
In [13]: np.object_(["hello world"]).real
Out[13]: array(['hello world'], dtype=object)

In [14]: np.object_(["hello world"]).imag
Out[14]: array([0], dtype=object)
```

Reply:
Issue opened at https://github.com/numpy/numpy/issues/10818